lib/elasticsearch.js

```diff
@@ -0,0 +1,540 @@
+
+		this.config = {
+			protocol: config.protocol || '',
+			host: config.host || ES_DEFAULT_HOST,
+			port: config.port || '',
+			user: config.user || '',
+			password: config.password || '',
+			limit: config.limit || DEFAULT_LIMIT,
+			awsCredentials: config.awsCredentials || false
+		};
+
+		this.config.host = this.config.host.replace(this.config.protocol, '');
+	}
+
+	/**
+	 * Elasticsearch connection URL prefix
+	 * @returns {String} Elasticsearch URL prefix
+	*/
+	get userPrefix() {
+		return this.config.user ? `${this.config.user}:${this.config.password}@` : '';
+	}
+
+	/**
+	 * Checks the connection to the Elasticsearch Client
+	 * @returns {elasticsearch.Client} Elasticsearch client with the connection
+	 * @throws if the connection is not successfull
+	*/
+	getClient() {
+		if(!this.client) {
+			try {
+				this.client = new elasticsearch.Client({
+					node: `${this.config.protocol}${this.userPrefix}${this.config.host}${this.config.port ? `:${this.config.port}` : ''}`,
+					...this.config.awsCredentials ? { connectionClass: AWSESConnection, amazonES: { credentials: new AWS.EnvironmentCredentials('AWS') } } : {}
+				});
+			} catch(err) {
+				throw new ElasticSearchError(err.message, ElasticSearchError.codes.ELASTICSEARCH_ERROR);
+			}
+		}
+		return this.client;
+	}
+
+	/**
+	 * Validates the model structure
+	 * @param {Model} model Model instance
+	 * @throws If the model is invalid
+	 */
+	validateModel(model) {
+
+		if(!model)
+			throw new ElasticSearchError('Empty model', ElasticSearchError.codes.INVALID_MODEL);
+
+		if(!model.constructor.table || !model.constructor.fields)
+			throw new ElasticSearchError('Invalid model', ElasticSearchError.codes.INVALID_MODEL);
+
+		if(!Array.isArray(model.constructor.fields))
+			throw new ElasticSearchError('Invalid fields. it must be an array, see model.constructor.fields.', ElasticSearchError.codes.INVALID_MODEL);
+
+	}
+
+	/**
+	 * Build elasticsearch index with mappings for sortings
+	 * @param {Model} model Model instance
+	 */
+	async buildIndex(model) {
+
+		this.validateModel(model);
+
+		try {
+
+			const client = this.getClient();
+			const exists = await client.indices.exists({ index: model.constructor.table });
+
+			if(!exists.body)
+				await client.indices.create({ index: model.constructor.table });
+
+			await client.indices.putMapping({
+				index: model.constructor.table,
+				type: '_doc',
+				body: this.getMappingsFromModel(model)
+			});
+
+		} catch(err) {
+			throw new ElasticSearchError(err.message, ElasticSearchError.codes.ELASTICSEARCH_ERROR);
+		}
+	}
+
+	/**
+	 * Analyses the model fields and build the mappings query for elasticsearch
+	 * @param {Model} model Model instance
+	 * @returns {Object} elasticsearch mappings query
+	 */
+	getMappingsFromModel(model) {
+
+		this.validateModel(model);
+
+		const parsedMappings = { properties: {} };
+
+		for(const field of model.constructor.fields) {
+			if(typeof field === 'object' && !Array.isArray(field)) {
```

Comment:
La intención es que si no es un objeto que tenga el `type` definido se tome como *default* qu e sea `string`.
Pero si `field` es un `Array` no entra en el condicional y lo tratará de setear como `string` y va a crashear.
```suggestion
			if(Array.isArray(field))
				continue;
			if(typeof field === 'object') {
```